test/scale/scale_tester.go

```diff
@@ -0,0 +1,292 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scale
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/onsi/ginkgo"
+	scale "k8s.io/api/autoscaling/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"sigs.k8s.io/azuredisk-csi-driver/test/e2e/driver"
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+	nodeutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/node"
+	podutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/pod"
+	volutil "sigs.k8s.io/azuredisk-csi-driver/test/utils/volume"
+)
+
+type PodSchedulingWithPVScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+type PodSchedulingOnFailoverScaleTest struct {
+	CSIDriver              driver.DynamicPVTestDriver
+	Pod                    testtypes.PodDetails
+	PodCount               int
+	Replicas               int
+	StorageClassParameters map[string]string
+}
+
+func (t *PodSchedulingWithPVScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	totalNumberOfPods := t.Replicas
+	var oneMinPodsRunning,
+		twoMinPodsRunning,
+		threeMinPodsRunning,
+		fiveMinPodsRunning,
+		tenMinPodsRunning = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+	var oneMinPodsDeleted,
+		twoMinPodsDeleted,
+		threeMinPodsDeleted,
+		fiveMinPodsDeleted,
+		tenMinPodsDeleted = totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods, totalNumberOfPods
+
+	startScaleOut := time.Now()
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+	statefulSet, cleanupStatefulSet := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, t.Replicas, t.StorageClassParameters, &tStorageClass)
+	for i := range cleanupStatefulSet {
+		i := i
+		defer cleanupStatefulSet[i]()
+	}
+	statefulSet.CreateWithoutWaiting()
+	numberOfRunningPods := 0
+	ticker := time.NewTicker(time.Minute)
+	tickerStop := make(chan bool)
+	go func() {
+		time.Sleep(11 * time.Minute)
+		tickerStop <- true
+	}()
+	tickerCount := 0
+	stopLoop := false
+	for {
+		select {
+		case <-tickerStop:
+			stopLoop = true
+
+		case <-ticker.C:
+			tickerCount = tickerCount + 1
+			numberOfRunningPods = podutil.CountAllPodsWithMatchingLabel(client, namespace, map[string]string{"app": "azuredisk-volume-tester"}, string(v1.PodRunning))
+			if numberOfRunningPods >= totalNumberOfPods {
+				stopLoop = true
+			}
+
+			if tickerCount == 1 {
```

Comment:
instead of using `tickerCount`, we can calculate the timestamp of each of these checks based on when the test starts and compare it to current time in ticker.C. Then since we want to get the exact number of pods during that time, we have to issue a `List` request when the condition is satisfied.

Reply:
With min as ticker interval, we are fine and don't need to do anything here.